src/librustc_trans/trans/context.rs

```diff
@@ -977,11 +942,62 @@ fn declare_intrinsic(ccx: &CrateContext, key: &str) -> Option<ValueRef> {
                 ccx.intrinsics().borrow_mut().insert($name, f.clone());
                 return Some(f);
             }
+        );
+        ($name:expr, $cname:ident ($($arg:expr),*) -> $ret:expr) => (
+            if !is_like_pnacl {
+                // The `if key == $name` is already in ifn!
+                ifn!($name, fn($($arg),*) -> $ret);
+            } else if key == $name {
+                let f = declare::declare_cfn(ccx, stringify!($cname),
+                                             Type::func(&[$($arg),*], &$ret),
+                                             ccx.tcx().mk_nil());
+                ccx.intrinsics().borrow_mut().insert($name, f.clone());
+                return Some(f);
+            }
         )
     }
 
     compatible_ifn!("llvm.assume", noop(llvmcompat_assume(i1) -> void), 6);
 
+    // These must be redirected to `libm` here so PNaCl's linker will link in
+    // these functions from `libm`.
+    compatible_ifn!("llvm.copysign.f32", copysignf(t_f32, t_f32) -> t_f32);
```

Comment:
@kripken

> Yes, but that's a bug in Emscripten :) That particular intrinsic (llvm.powi.*) was not implemented. Fixed now in kripken/emscripten-fastcomp@fd10012

Ah, that's fair, but I was speaking generally. The majority of LLVM's math intrinsics are still not handled; only calls to the libm versions are "handled" (and that's just an optimization).

Perhaps there's something I'm missing? Checking with `em++ -emit-llvm` gives me (using a variation of my previous example, with `__builtin_sinf`):

```llvm
; ModuleID = 'test.ll'
target datalayout = "e-p:32:32-i64:64-v128:32:128-n32-S128"
target triple = "asmjs-unknown-emscripten"

@.str = private unnamed_addr constant [7 x i8] c"v == 0\00", align 1
@.str1 = private unnamed_addr constant [7 x i8] c"test.c\00", align 1
@__func__.main = private unnamed_addr constant [5 x i8] c"main\00", align 1

define i32 @main() #0 {
  %1 = alloca i32, align 4
  %v = alloca float, align 4
  store i32 0, i32* %1
  %2 = call float @sinf(float 0x40091EB860000000) #3
  store float %2, float* %v, align 4
  %3 = load float, float* %v, align 4
  %4 = fcmp oeq float %3, 0.000000e+00
  br i1 %4, label %7, label %5

; <label>:5                                       ; preds = %0
  call void @__assert_fail(i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str, i32 0, i32 0), i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str1, i32 0, i32 0), i32 6, i8* getelementptr inbounds ([5 x i8], [5 x i8]* @__func__.main, i32 0, i32 0)) #4
  unreachable
                                                  ; No predecessors!
  br label %7

; <label>:7                                       ; preds = %6, %0
  %8 = phi i1 [ true, %0 ], [ false, %6 ]
  ret i32 0
}

; Function Attrs: nounwind readnone
declare float @sinf(float) #1

; Function Attrs: noreturn
declare void @__assert_fail(i8*, i8*, i32, i8*) #2

attributes #0 = { "less-precise-fpmad"="false" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "stack-protector-buffer-size"="8" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #1 = { nounwind readnone "less-precise-fpmad"="false" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "stack-protector-buffer-size"="8" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #2 = { noreturn "less-precise-fpmad"="false" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-nans-fp-math"="false" "stack-protector-buffer-size"="8" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #3 = { nounwind readnone }
attributes #4 = { noreturn }

!llvm.ident = !{!0}

!0 = !{!"clang version 3.7.0 "}
```

So first off, it seems even calling `__builtin_sinf` doesn't use the LLVM `sin.f32` intrinsic.

I went ahead and checked by changing the `sinf` to `llvm.sin.f32` in the above IR and fed it to `em++`: I got "warning: unresolved symbol: llvm_sin_f32". So JS really doesn't handle all of the math-y intrinsics.

> As I understand it, the PNaCl ABI is a definition of the PNaCl distribution format, i.e., the thing that ships to browsers. It makes sense to limit that, so that PNaCl executables are as portable as possible. However, there isn't a reason I am aware of to not support all LLVM intrinsics in the PNaCl developer-side compiler. If you don't handle llvm.powi.* and all other math intrinsics there, then you need hacks in every single compiler targeting PNaCl.

Again, JS codegen doesn't handle the other math intrinsics either; as you've said, it hasn't yet been a problem for the other languages targeting JS, meaning the compilers employed hacks like the one being discussed here or they never used the intrinsics in the first place, opting to always call into libm.

Having said all that, I think I can fix this for both projects. See https://github.com/rust-lang/rust/pull/28355#issuecomment-144577499.